File: JobTracker/Services/ApplicationMatcher.swift
//
//  ApplicationMatcher.swift
//  JobTracker
//
//  Groups classified emails into job applications using layered signals:
//
//    1. Gmail thread id                     → certain match  (1.0)
//    2. same company + similar role         → strong match   (0.9)
//    3. same company + partial role overlap → probable       (0.75, review)
//    4. same company + a missing role,
//       close in time                       → probable       (0.7, review)
//    5. known sender address                → weak            (0.6, review)
//
//  Role comparison is token-based (season/year tokens stripped) so
//  "Software Engineer Intern" matches "Software Engineering Intern —
//  Summer 2026". Company keys match exactly or by whole-token containment
//  ("amazon" ⊂ "amazon web services").
//
//  Indexes are built once per sync session and updated incrementally —
//  including thread ids attached to existing applications mid-run, so the
//  second email of a thread always matches at 1.0.
//

import Foundation

@MainActor
final class ApplicationMatcher {

    struct Match {
        let application: JobApplication
        let confidence: Double
        var needsReview: Bool { confidence < 0.9 }
    }

    /// How close in time a role-less email must be to an application with
    /// the same company to be considered the same process.
    private static let proximityWindow: TimeInterval = 120 * 24 * 3600

    private var byThread: [String: JobApplication] = [:]
    private var byCompany: [String: [JobApplication]] = [:]
    private var bySender: [String: JobApplication] = [:]

    init(existing: [JobApplication]) {
        for application in existing { register(application) }
    }

    /// Adds a new application to the session indexes.
    func register(_ application: JobApplication) {
        for threadId in application.threadIds where !threadId.isEmpty {
            byThread[threadId] = application
        }
        if !application.companyKey.isEmpty {
            byCompany[application.companyKey, default: []].append(application)
        }
        if let contact = application.contactEmail.flatMap(Self.emailAddress) {
            bySender[contact] = application
        }
    }

    /// Records that `threadId` now belongs to `application` — called
    /// whenever the pipeline appends a thread to an existing application so
    /// follow-ups in that thread match at 1.0 instead of re-running fuzzy
    /// matching (the old behavior caused duplicates).
    func associate(threadId: String, with application: JobApplication) {
        guard !threadId.isEmpty else { return }
        byThread[threadId] = application
    }

    // MARK: - Matching

    func match(message: FetchedMessage, result: ClassificationResult) -> Match? {
        // 1. Same Gmail thread — same application, always.
        if !message.threadId.isEmpty, let app = byThread[message.threadId] {
            return Match(application: app, confidence: 1.0)
        }

        let companyKey = JobApplication.normalizeCompany(result.company)
        let candidates = companyCandidates(for: companyKey)

        if !candidates.apps.isEmpty {
            let incomingTokens = Self.roleTokens(result.role)

            // 2./3. Role-similarity ranking within the company.
            var bestSimilar: (app: JobApplication, similarity: Double)?
            var windowCandidates: [JobApplication] = []
            for candidate in candidates.apps {
                let candidateTokens = Self.roleTokens(candidate.roleTitle)
                if incomingTokens.isEmpty || candidateTokens.isEmpty {
                    // 4. A side lacks a role: candidate only if close in time.
                    let anchor = candidate.lastEmailDate ?? candidate.lastUpdated
                    if abs(anchor.timeIntervalSince(message.date)) < Self.proximityWindow {
                        windowCandidates.append(candidate)
                    }
                    continue
                }
                let similarity = Self.similarity(incomingTokens, candidateTokens)
                if similarity > (bestSimilar?.similarity ?? 0) {
                    bestSimilar = (candidate, similarity)
                }
            }

            if let best = bestSimilar {
                if best.similarity >= 0.6 {
                    // Same role in different words. Exact-company gets full
                    // strength; containment-matched company keeps review.
                    return Match(application: best.app,
                                 confidence: candidates.exactCompany ? 0.9 : 0.75)
                }
                if best.similarity >= 0.35 {
                    return Match(application: best.app, confidence: 0.75)
                }
            }

            // Role-less matching must respect multiple parallel applications
            // at the same company (e.g. several roles applied the same day):
            // unambiguous single candidate attaches; multiple candidates
            // attach to the nearest-in-time but flagged for review so the
            // user can detach if the guess is wrong.
            if incomingTokens.isEmpty || candidates.apps.allSatisfy({ Self.roleTokens($0.roleTitle).isEmpty }) {
                if windowCandidates.count == 1 {
                    return Match(application: windowCandidates[0], confidence: 0.7)
                }
                if windowCandidates.count > 1 {
                    let nearest = windowCandidates.min { a, b in
                        let anchorA = a.lastEmailDate ?? a.lastUpdated
                        let anchorB = b.lastEmailDate ?? b.lastUpdated
                        return abs(anchorA.timeIntervalSince(message.date)) <
                               abs(anchorB.timeIntervalSince(message.date))
                    }
                    if let nearest {
                        return Match(application: nearest, confidence: 0.6)
                    }
                }
            }
            // Distinct roles at the same company → genuinely separate
            // applications; fall through to create a new one.
        }

        // 5. Known recruiter/sender address, when the model found no company.
        if companyKey.isEmpty,
           let sender = Self.emailAddress(message.sender),
           let app = bySender[sender] {
            return Match(application: app, confidence: 0.6)
        }

        return nil
    }

    /// Applications at the same company: exact key match, else whole-token
    /// containment ("amazon" ⊂ "amazon web services", either direction).
    private func companyCandidates(for key: String) -> (apps: [JobApplication], exactCompany: Bool) {
        guard !key.isEmpty else { return ([], false) }
        if let exact = byCompany[key] { return (exact, true) }

        let tokens = Set(key.split(separator: " ").map(String.init))
        var contained: [JobApplication] = []
        for (otherKey, apps) in byCompany {
            let otherTokens = Set(otherKey.split(separator: " ").map(String.init))
            if tokens.isSubset(of: otherTokens) || otherTokens.isSubset(of: tokens) {
                contained.append(contentsOf: apps)
            }
        }
        return (contained, false)
    }

    // MARK: - Role similarity

    /// Tokens that describe the cycle, not the role, and connective noise.
    /// NOTE: "intern"/"grad" are deliberately KEPT — an internship and a
    /// full-time role at the same company are different applications.
    private static let noiseTokens: Set<String> = [
        "summer", "fall", "autumn", "winter", "spring",
        "the", "a", "an", "of", "and", "for", "at", "in", "position", "role"
    ]

    /// Lowercased word tokens with cycle years/seasons and noise removed.
    nonisolated static func roleTokens(_ role: String) -> Set<String> {
        Set(
            role.lowercased()
                .components(separatedBy: CharacterSet.alphanumerics.inverted)
                .filter { !$0.isEmpty }
                .filter { !noiseTokens.contains($0) }
                .filter { !($0.count == 4 && $0.hasPrefix("20") && Int($0) != nil) }
                .map { token in
                    // Light stemming so "engineering" matches "engineer".
                    token.hasSuffix("ing") && token.count > 6
                        ? String(token.dropLast(3)) : token
                }
        )
    }

    /// Jaccard similarity of two token sets.
    nonisolated static func similarity(_ a: Set<String>, _ b: Set<String>) -> Double {
        guard !a.isEmpty, !b.isEmpty else { return 0 }
        let intersection = a.intersection(b).count
        let union = a.union(b).count
        return Double(intersection) / Double(union)
    }

    // MARK: - Sender helpers

    /// Extracts "user@host" from a From header like "Name <user@host>".
    nonisolated static func emailAddress(_ from: String) -> String? {
        if let open = from.lastIndex(of: "<"), let close = from.lastIndex(of: ">"),
           open < close {
            return String(from[from.index(after: open)..<close]).lowercased()
        }
        let trimmed = from.trimmingCharacters(in: .whitespaces).lowercased()
        return trimmed.contains("@") ? trimmed : nil
    }

    /// Mail providers and ATS platforms whose domain never names the
    /// hiring company.
    private static let genericDomains: Set<String> = [
        "gmail", "googlemail", "outlook", "hotmail", "yahoo", "icloud", "aol",
        "greenhouse", "lever", "myworkday", "workday", "icims",
        "smartrecruiters", "ashbyhq", "jobvite", "taleo", "successfactors",
        "bamboohr", "workable", "linkedin", "indeed", "ziprecruiter",
        "hackerrank", "codesignal", "hirevue", "karat", "codility",
        "workablemail", "candidates", "notifications", "mail", "email", "no-reply"
    ]

    /// Best-effort company name from a sender address, e.g.
    /// "careers@stripe.com" → "Stripe". Nil for generic/ATS domains.
    nonisolated static func companyFromSender(_ from: String) -> String? {
        guard let address = emailAddress(from),
              let atIndex = address.firstIndex(of: "@") else { return nil }
        let host = String(address[address.index(after: atIndex)...])
        let labels = host.split(separator: ".").map(String.init)
        // Take the label left of the TLD ("careers.tiktok.com" → "tiktok").
        guard labels.count >= 2 else { return nil }
        let label = labels[labels.count - 2]
        guard label.count > 1, !genericDomains.contains(label),
              !labels.contains(where: { genericDomains.contains($0) }) else { return nil }
        return label.prefix(1).uppercased() + label.dropFirst()
    }
}

File: windows/JobTracker/Services/ApplicationMatcher.cs
// Groups classified emails into job applications using layered signals:
//
//   1. Gmail thread id                     → certain match  (1.0)
//   2. same company + similar role         → strong match   (0.9)
//   3. same company + partial role overlap → probable       (0.75, review)
//   4. same company + a missing role,
//      close in time                       → probable       (0.7, review)
//   5. known sender address                → weak            (0.6, review)
//
// Role comparison is token-based (season/year tokens stripped) so
// "Software Engineer Intern" matches "Software Engineering Intern —
// Summer 2026". Company keys match exactly or by whole-token containment
// ("amazon" ⊂ "amazon web services").
//
// Indexes are built once per sync session and updated incrementally —
// including thread ids attached to existing applications mid-run, so the
// second email of a thread always matches at 1.0.

using System.Text.RegularExpressions;
using JobTracker.Models;

namespace JobTracker.Services;

public sealed partial class ApplicationMatcher
{
    public sealed record Match(JobApplication Application, double Confidence)
    {
        public bool NeedsReview => Confidence < 0.9;
    }

    /// How close in time a role-less email must be to an application with
    /// the same company to be considered the same process.
    private static readonly TimeSpan ProximityWindow = TimeSpan.FromDays(120);

    private readonly Dictionary<string, JobApplication> _byThread = [];
    private readonly Dictionary<string, List<JobApplication>> _byCompany = [];
    private readonly Dictionary<string, JobApplication> _bySender = [];

    public ApplicationMatcher(IEnumerable<JobApplication> existing)
    {
        foreach (var application in existing) Register(application);
    }

    /// Adds a new application to the session indexes.
    public void Register(JobApplication application)
    {
        foreach (var threadId in application.ThreadIds.Where(t => !string.IsNullOrEmpty(t)))
        {
            _byThread[threadId] = application;
        }
        if (!string.IsNullOrEmpty(application.CompanyKey))
        {
            if (!_byCompany.TryGetValue(application.CompanyKey, out var list))
            {
                list = [];
                _byCompany[application.CompanyKey] = list;
            }
            list.Add(application);
        }
        var contact = application.ContactEmail is not null ? EmailAddress(application.ContactEmail) : null;
        if (contact is not null) _bySender[contact] = application;
    }

    /// Records that `threadId` now belongs to `application` — called
    /// whenever the pipeline appends a thread to an existing application so
    /// follow-ups in that thread match at 1.0 instead of re-running fuzzy
    /// matching (the old behavior caused duplicates).
    public void Associate(string threadId, JobApplication application)
    {
        if (string.IsNullOrEmpty(threadId)) return;
        _byThread[threadId] = application;
    }

    // MARK: - Matching

    public Match? MatchMessage(FetchedMessage message, ClassificationResult result)
    {
        // 1. Same Gmail thread — same application, always.
        if (!string.IsNullOrEmpty(message.ThreadId) && _byThread.TryGetValue(message.ThreadId, out var byThread))
        {
            return new Match(byThread, 1.0);
        }

        var companyKey = JobApplication.NormalizeCompany(result.Company);
        var (candidates, exactCompany) = CompanyCandidates(companyKey);

        if (candidates.Count > 0)
        {
            var incomingTokens = RoleTokens(result.Role);

            // 2./3. Role-similarity ranking within the company.
            (JobApplication App, double Similarity)? bestSimilar = null;
            List<JobApplication> windowCandidates = [];
            foreach (var candidate in candidates)
            {
                var candidateTokens = RoleTokens(candidate.RoleTitle);
                if (incomingTokens.Count == 0 || candidateTokens.Count == 0)
                {
                    // 4. A side lacks a role: candidate only if close in time.
                    var anchor = candidate.LastEmailDate ?? candidate.LastUpdated;
                    if ((anchor - message.Date).Duration() < ProximityWindow)
                    {
                        windowCandidates.Add(candidate);
                    }
                    continue;
                }
                var similarity = Similarity(incomingTokens, candidateTokens);
                if (similarity > (bestSimilar?.Similarity ?? 0))
                {
                    bestSimilar = (candidate, similarity);
                }
            }

            if (bestSimilar is { } best)
            {
                if (best.Similarity >= 0.6)
                {
                    // Same role in different words. Exact-company gets full
                    // strength; containment-matched company keeps review.
                    return new Match(best.App, exactCompany ? 0.9 : 0.75);
                }
                if (best.Similarity >= 0.35)
                {
                    return new Match(best.App, 0.75);
                }
            }

            // Role-less matching must respect multiple parallel applications
            // at the same company (e.g. several roles applied the same day):
            // unambiguous single candidate attaches; multiple candidates
            // attach to the nearest-in-time but flagged for review so the
            // user can detach if the guess is wrong.
            if (incomingTokens.Count == 0 || candidates.All(c => RoleTokens(c.RoleTitle).Count == 0))
            {
                if (windowCandidates.Count == 1)
                {
                    return new Match(windowCandidates[0], 0.7);
                }
                if (windowCandidates.Count > 1)
                {
                    var nearest = windowCandidates.MinBy(c =>
                    {
                        var anchor = c.LastEmailDate ?? c.LastUpdated;
                        return (anchor - message.Date).Duration();
                    });
                    if (nearest is not null) return new Match(nearest, 0.6);
                }
            }
            // Distinct roles at the same company → genuinely separate
            // applications; fall through to create a new one.
        }

        // 5. Known recruiter/sender address, when the model found no company.
        if (companyKey.Length == 0)
        {
            var sender = EmailAddress(message.Sender);
            if (sender is not null && _bySender.TryGetValue(sender, out var bySender))
            {
                return new Match(bySender, 0.6);
            }
        }

        return null;
    }

    /// Applications at the same company: exact key match, else whole-token
    /// containment ("amazon" ⊂ "amazon web services", either direction).
    private (List<JobApplication> Apps, bool ExactCompany) CompanyCandidates(string key)
    {
        if (key.Length == 0) return ([], false);
        if (_byCompany.TryGetValue(key, out var exact)) return (exact, true);

        var tokens = key.Split(' ').ToHashSet();
        List<JobApplication> contained = [];
        foreach (var (otherKey, apps) in _byCompany)
        {
            var otherTokens = otherKey.Split(' ').ToHashSet();
            if (tokens.IsSubsetOf(otherTokens) || otherTokens.IsSubsetOf(tokens))
            {
                contained.AddRange(apps);
            }
        }
        return (contained, false);
    }

    // MARK: - Role similarity

    /// Tokens that describe the cycle, not the role, and connective noise.
    /// NOTE: "intern"/"grad" are deliberately KEPT — an internship and a
    /// full-time role at the same company are different applications.
    private static readonly HashSet<string> NoiseTokens =
    [
        "summer", "fall", "autumn", "winter", "spring",
        "the", "a", "an", "of", "and", "for", "at", "in", "position", "role",
    ];

    [GeneratedRegex(@"[^a-z0-9]+")]
    private static partial Regex NonAlphanumeric();

    /// Lowercased word tokens with cycle years/seasons and noise removed.
    public static HashSet<string> RoleTokens(string role)
    {
        var lower = role.ToLowerInvariant();
        var words = NonAlphanumeric().Split(lower).Where(w => w.Length > 0);
        return words
            .Where(w => !NoiseTokens.Contains(w))
            .Where(w => !(w.Length == 4 && w.StartsWith("20") && int.TryParse(w, out _)))
            .Select(w => w.EndsWith("ing") && w.Length > 6 ? w[..^3] : w)
            .ToHashSet();
    }

    /// Jaccard similarity of two token sets.
    public static double Similarity(HashSet<string> a, HashSet<string> b)
    {
        if (a.Count == 0 || b.Count == 0) return 0;
        var intersection = a.Intersect(b).Count();
        var union = a.Union(b).Count();
        return (double)intersection / union;
    }

    // MARK: - Sender helpers

    /// Extracts "user@host" from a From header like "Name <user@host>".
    public static string? EmailAddress(string from)
    {
        var open = from.LastIndexOf('<');
        var close = from.LastIndexOf('>');
        if (open >= 0 && close >= 0 && open < close)
        {
            return from[(open + 1)..close].ToLowerInvariant();
        }
        var trimmed = from.Trim().ToLowerInvariant();
        return trimmed.Contains('@') ? trimmed : null;
    }

    /// Mail providers and ATS platforms whose domain never names the
    /// hiring company.
    private static readonly HashSet<string> GenericDomains =
    [
        "gmail", "googlemail", "outlook", "hotmail", "yahoo", "icloud", "aol",
        "greenhouse", "lever", "myworkday", "workday", "icims",
        "smartrecruiters", "ashbyhq", "jobvite", "taleo", "successfactors",
        "bamboohr", "workable", "linkedin", "indeed", "ziprecruiter",
        "hackerrank", "codesignal", "hirevue", "karat", "codility",
        "workablemail", "candidates", "notifications", "mail", "email", "no-reply",
    ];

    /// Best-effort company name from a sender address, e.g.
    /// "careers@stripe.com" → "Stripe". Null for generic/ATS domains.
    public static string? CompanyFromSender(string from)
    {
        var address = EmailAddress(from);
        if (address is null) return null;
        var atIndex = address.IndexOf('@');
        if (atIndex < 0) return null;
        var host = address[(atIndex + 1)..];
        var labels = host.Split('.').Where(l => l.Length > 0).ToList();
        if (labels.Count < 2) return null;
        var label = labels[^2];
        if (label.Length <= 1 || GenericDomains.Contains(label) || labels.Any(GenericDomains.Contains))
        {
            return null;
        }
        return char.ToUpperInvariant(label[0]) + label[1..];
    }
}
